subprojects/frameworklauncher/yarn/src/main/java/com/microsoft/frameworklauncher/common/exit/FrameworkExitSpec.java

```diff
@@ -0,0 +1,334 @@
+        "Failed to deserialize AMDiagnostics from YARN, maybe it is truncated due to it is too long"));
+    /// Service External Error is already handled by host
+  }
+
+  private static UserContainerExitSpec normalize(UserContainerExitSpec userContainerExitSpec) {
+    Map<Integer, UserContainerExitInfo> normalizedSpec = new HashMap<>();
+
+    if (userContainerExitSpec != null && userContainerExitSpec.getSpec() != null) {
+      for (UserContainerExitInfo exitInfo : userContainerExitSpec.getSpec()) {
+        if (exitInfo.getCode() != null &&
+            exitInfo.getCode() >= 1 &&
+            exitInfo.getCode() <= 255 &&
+            exitInfo.getType() != ExitType.SUCCEEDED) {
+          normalizedSpec.put(exitInfo.getCode(), UserContainerExitInfo.newInstance(
+              exitInfo.getCode(),
+              exitInfo.getType() != null ? exitInfo.getType() : ExitType.UNKNOWN,
+              exitInfo.getDescription()));
+        }
+      }
+    }
+
+    return UserContainerExitSpec.newInstance(new ArrayList<>(normalizedSpec.values()));
+  }
+
+  public static UserContainerExitSpec initialize(UserContainerExitSpec userContainerExitSpec) {
+    LOGGER.logInfo("Initializing FrameworkExitSpec: Loading UserContainerExitSpec");
+
+    UserContainerExitSpec effectiveUserContainerExitSpec = normalize(userContainerExitSpec);
+    for (UserContainerExitInfo exitInfo : effectiveUserContainerExitSpec.getSpec()) {
+      spec.put(exitInfo.getCode(), new FrameworkExitInfo(
+          exitInfo.getCode(),
+          null,
+          exitInfo.getType(),
+          exitInfo.getDescription()));
+    }
+
+    return effectiveUserContainerExitSpec;
+  }
+
+  public static int lookupExitCode(int rawCode, String rawDiagnostics) {
+    if (rawDiagnostics == null) {
+      rawDiagnostics = "";
+    }
+
+    // Prefer to pick smaller ExitCode if multiple ExitCodes are recognized.
+    for (Map.Entry<Integer, FrameworkExitInfo> exitInfoKV : spec.entrySet()) {
+      int code = exitInfoKV.getKey();
+      FrameworkExitInfo info = exitInfoKV.getValue();
+
+      // Match RawCodePattern
+      if (rawCode != info.getRawCodePattern()) {
+        continue;
+      }
+
+      // Match RawDiagnosticsPattern
+      if (info.getRawDiagnosticsPattern() != null) {
+        if (!rawDiagnostics.toLowerCase().contains(
+            info.getRawDiagnosticsPattern().toLowerCase())) {
+          continue;
+        }
+      }
+
+      // Found Known ExitCode
+      return code;
+    }
+
+    // Preserve Unknown RawExitCode as ExitCode
+    return rawCode;
+  }
+
+  public static FrameworkExitInfo getExitInfo(int code) {
+    FrameworkExitInfo info = spec.get(code);
+    if (info != null) {
+      return info;
+    } else {
+      // Unknown FrameworkExitInfo
+      if (code >= 0) {
+        return new FrameworkExitInfo(
+            code,
+            null,
+            ExitType.UNKNOWN,
+            "Container exited with unknown exitcode which is issued from itself");
+      } else {
```

Comment:
ditto

Reply:
Why?

Reply:
It is for functional style for better understanding. Anyway it is code style, we can omit it in PR.

Reply:
I'm fine with it, just prefer "less nesting" code style. You can refer to https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting if you are interested.

Anyway, you can omit following comments for the code style

Reply:
Thanks for the info ! And since it is not deep, I will keep orignal style.